activerecord/lib/active_record/associations/preloader/through_association.rb

```diff
@@ -10,10 +10,13 @@ def preloaded_records
 
         def records_by_owner
           return @records_by_owner if defined?(@records_by_owner)
-          source_records_by_owner = source_preloaders.map(&:records_by_owner).reduce(:merge)
```

Comment:
We want to skip calling `records_by_owner` if everything is already loaded, otherwise we get an extra query in some of the nested through association tests.